File: src/js/4_BehavioralPatterns/6_Memento/memento_exercise.js
class Token
{
  constructor(value=0)
  {
    this.value = value;
  }
}

class Memento
{
  constructor()
  {
    this.tokens = [];
  }
}

class TokenMachine
{
  constructor()
  {
    this.tokens = [];
  }

  addTokenValue(value)
  {
    return this.addToken(new Token(value));
  }

  addToken(token)
  {
    this.tokens.push(token);
    let m = new Memento();
    m.tokens = this.tokens.map(t => new Token(t.value));
    return m;
  }

  revert(m)
  {
    this.tokens = m.tokens.map(t => new Token(t.value));
  }
}

describe('memento', function()
{
  it('single token test', function()
  {
    let tm = new TokenMachine();
    let m = tm.addTokenValue(123);
    tm.addTokenValue(456);
    tm.revert(m);
    expect(tm.tokens.length).toEqual(1);
    expect(tm.tokens[0].value).toEqual(123);
  });

  it('two token test', function() {
    let tm = new TokenMachine();
    tm.addTokenValue(1);
    let m = tm.addTokenValue(2);
    tm.addTokenValue(3);
    tm.revert(m);
    expect(tm.tokens.length).toEqual(2);
    expect(tm.tokens[0].value).toEqual(1);
    expect(tm.tokens[1].value).toEqual(2);
  });

  it('fiddle token test', function()
  {
    let tm = new TokenMachine();
    console.log('made a token with value 111 and kept a reference');
    let token = new Token(111);
    console.log('added this token to the list');
    tm.addToken(token);
    let m = tm.addTokenValue(222);
    console.log('changed this token\'s value to 333 :)');
    token.value = 333;
    tm.revert(m);

    expect(tm.tokens.length).toEqual(2);
    expect(tm.tokens[0].value).toEqual(111);
  });
});

File: src/java/4_BehavioralPatterns/6_Memento/MementoExercise.java
package com.activemesa.behavioral.memento.exercise;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

class Token
{
  public int value = 0;

  public Token(int value)
  {
    this.value = value;
  }
}

class Memento
{
  public List<Token> tokens = new ArrayList<>();
}

class TokenMachine
{
  public List<Token> tokens = new ArrayList<>();

  public Memento addToken(int value)
  {
    return addToken(new Token(value));
  }

  public Memento addToken(Token token)
  {
    tokens.add(token);
    Memento m = new Memento();
    m.tokens = tokens.stream()
      .map(t -> new Token(t.value))
      .collect(Collectors.toList());
    return m;
  }

  public void revert(Memento m)
  {
    tokens = m.tokens.stream()
      .map(t -> new Token(t.value))
      .collect(Collectors.toList());
  }
}